### PR TITLE
Bugfix: Standalone dotlambda just before EOF

### PR DESCRIPTION
**Pull request by T-Gro**
This addresses https://github.com/dotnet/fsharp/issues/16220 .

If my hypothesis is correct, the difference was caused by an end-of-file marker.
In a bigger statement (e.g. with a "let" binding), rules higher up cover for an EOF.
In a standalone parsing of just the dotlambda construct, this is not covered.

Why did matter? Because EOF is one possibility for "recover", which was meant for error handling when this feature was being implemented. But EOF does not always mean an error.

@auduchinok  , @DedSec256  :
I did not like the fact of special casing EOF, because most other constructs do not need it either, it is dealt with higher up.
I would more than welcome any suggestion of yours, both for the EOF as well deduplicating the rules here.

### CODE REVIEW

**Review by auduchinok — COMMENTED** [review]
@T-Gro I see two problems with the existing rules:
* Why do the rules use different rules for expression, i.e. `atomicExpr` vs `appExpr`?
  `atomicExpr` covers the cases with high-precedence applications, so `_.M()` is already covered. `appExpr` allows `_.M ()` to be parsed but adds an additional error. It looks wrong to me, we should only parse atomic expressions here.
* Why does the second rule have `recover` at all? It doesn't look like there's something is missing in it where we expect the recovery to kick in.

It seems that simply removing the second rule can fix it (it looks too wrong and interferes with the first rule), while adding this third one only makes the problem deeper.

That will allow `_.P 123` as:
```
appExpr
  dotLambdaExpr
    _
    .
    namedExpr
      P
  constExpr
    123
```
and `f _.P 123` to be parsed as:
```
appExpr
  appExpr
    namedExpr
      f
    dotLambdaExpr
      _
      .
      namedExpr
        P
  constExpr
    123
```

File: tests/service/data/SyntaxTree/DotLambda/TopLevelLet.fs [hunk 1]
```diff
@@ -0,0 +1 @@
+let _ = 1 |> _.ToString()
```
[the remark this answers is not in the document]
Comment (T-Gro) [reply tests/service/data/SyntaxTree/DotLambda/TopLevelLet.fs:1]: Folders actually do help me with filtering tests via cmd invocations, and I do not need to rely on file naming being typed exactly the same all the time.

I can move as a subfolder into the expression folder, that works.

**Comment by auduchinok** [comment]
> The second rule only exists in order to give a meaningful error message instead of a generic one, while keeping some parser recovery working.

Could you please show an example of the generic error message when this rule is removed? I think it's very important to fix the atomic expressions non being atomic here, and the only way I see is to remove this rule.

File: tests/service/data/SyntaxTree/DotLambda/UnderscoreToString.fs.bsl [hunk 2]
```diff
@@ -14,5 +14,3 @@ ImplFile
           { LeadingKeyword = None })], (true, true),
       { ConditionalDirectives = []
         CodeComments = [] }, set []))
-
-(1,0)-(1,1) parse error  _. shorthand syntax for lambda functions can only be used with atomic expressions. That means expressions with no whitespace unless enclosed in parentheses.
```
Comment (T-Gro) [inline tests/service/data/SyntaxTree/DotLambda/UnderscoreToString.fs.bsl:18]: Getting rid of this error is a good thing, this was just `_.ToString()`.

File: tests/service/data/SyntaxTree/DotLambda/UnderscoreToFunctionNallWithSpaceAndUnitApplication.fs.bsl [hunk 3]
```diff
@@ -17,18 +17,16 @@ ImplFile
...
+                     (1,13--1,30)), (1,4--1,10), Yes (1,0--1,30),
                   { LeadingKeyword = Let (1,0--1,3)
                     InlineKeyword = None
                     EqualsRange = Some (1,11--1,12) })], (1,0--1,30))],
           PreXmlDocEmpty, [], None, (1,0--1,30), { LeadingKeyword = None })],
       (true, true), { ConditionalDirectives = []
                       CodeComments = [] }, set []))
-
-(1,0)-(1,30) parse error Incomplete structured construct at or before this point in expression
-(1,13)-(1,14) parse error  _. shorthand syntax for lambda functions can only be used with atomic expressions. That means expressions with no whitespace unless enclosed in parentheses.
```
Comment (T-Gro) [inline tests/service/data/SyntaxTree/DotLambda/UnderscoreToFunctionNallWithSpaceAndUnitApplication.fs.bsl:34]: This is a nonatomic `let myFunc = _.MyMethodCall ()`

It was previously failing with the error message in the diff, and now it passes.
I guess I could move the error into typechecking, if keeping things in the parser does not make sense.
(even though I think keeping it in the parser should be doable?)

File: tests/service/data/SyntaxTree/DotLambda/WithNonTupledFunctionCall.fs.bsl [hunk 4]
```diff
@@ -14,23 +14,21 @@ ImplFile
...
+                     (1,13--1,37)), (1,4--1,10), Yes (1,0--1,37),
                   { LeadingKeyword = Let (1,0--1,3)
                     InlineKeyword = None
                     EqualsRange = Some (1,11--1,12) })], (1,0--1,37))],
           PreXmlDocEmpty, [], None, (1,0--1,37), { LeadingKeyword = None })],
       (true, true), { ConditionalDirectives = []
                       CodeComments = [] }, set []))
-
-(1,0)-(1,37) parse error Incomplete structured construct at or before this point in expression
-(1,13)-(1,14) parse error  _. shorthand syntax for lambda functions can only be used with atomic expressions. That means expressions with no whitespace unless enclosed in parentheses.
```
Comment (T-Gro) [inline tests/service/data/SyntaxTree/DotLambda/WithNonTupledFunctionCall.fs.bsl:36]: This code was `let myFunc = _.ThisIsMyFunction a b c`, which we also want to keep failing.

**Comment by T-Gro** [comment]
> > The second rule only exists in order to give a meaningful error message instead of a generic one, while keeping some parser recovery working.
> 
> Could you please show an example of the generic error message when this rule is removed? I think it's very important to fix the atomic expressions non being atomic here, and the only way I see is to remove this rule.

It is now visible in the diff + I added commented explanations.
I see three options:
- I move the "nonatomic" error into typechecker
- I keep 1 rule only in the parser, but implemented addition logic inside the rule to separate atomic/nonatomic
- I revert to two rules, one passing (atomic) and one failing (nonatomic), but will avoid using the 'recover' in the rule.

The second one looks best to me, WDYT?

**Comment by auduchinok** [comment]
@T-Gro Thanks, all the examples are good. Looking at these, I agree the second option is the best at things stand now. A simple check in the type checker could be checking that dot lambdas aren't allowed as function expressions in application expressions:

This is allowed, because `_.P` is the argument expression in the inner app expression:
```
f _.P 123
```
```
appExpr
  appExpr
    namedExpr
      f
    dotLambdaExpr
      _
      .
      namedExpr
        P
  constExpr
    123
```
And this is not allowed, because `_.P` is the function expr:
```
let _ = _.P 123
```
```
appExpr
  dotLambdaExpr
    _
    .
    namedExpr
      P
  constExpr
    123
```

Can the special precedence also be removed if non-atomic expressions are no longer allowed there?

**Review by auduchinok — APPROVED** [review]
It looks good to me now, it's good to have the underlying problem fixed!

**Comment by T-Gro** [comment]
Agree.

I also wanted to remove the %prec, tried that, but I started to get test failures which did not repro locally.
But it meant that _.Prop was getting parsed as an identifier and dot access, and not as a dot lambda, I wasn't sure why since I could not repro.

=> I will keep that outside of this PR, so that it can get it as it is.


== this is ready for reviews and merging it

File: src/Compiler/Checking/CheckExpressions.fs [hunk 5]
```diff
@@ -5308,6 +5308,10 @@ and TcExprThen (cenv: cenv) overallTy env tpenv isArg synExpr delayed =
     // f(x)  // hpa=true
     // f[x]  // hpa=true
     | SynExpr.App (hpa, isInfix, func, arg, mFuncAndArg) ->
+        match func with
+        | SynExpr.DotLambda _ -> errorR(Error(FSComp.SR.tcDotLambdaAtNotSupportedExpression(), func.Range))
```
Comment (abonie) [inline src/Compiler/Checking/CheckExpressions.fs:5312]: I am a bit confused, what case is this type checking? Something like `(_.ToString()) 1`?

Comment (T-Gro) [reply src/Compiler/Checking/CheckExpressions.fs:5312]: This for example:

let _ = _.PropertyWhichIsAFunction 123

Comment (T-Gro) [reply src/Compiler/Checking/CheckExpressions.fs:5312]: It is an application a function (left) and argument (right), separated with a space.

**Review by abonie — APPROVED** [review]

**Review by psfinaki — APPROVED** [review]

**Pull request merged**